x-pack/plugins/apm/server/routes/environments/get_environments.ts

```diff
@@ -39,44 +39,29 @@ export async function getEnvironments({
     ? 'get_environments_for_service'
     : 'get_environments';
 
-  const params = {
+  const params: APMEventESTermsEnumRequest = {
     apm: {
       events: [
         getProcessorEventForTransactions(searchAggregatedTransactions),
         ProcessorEvent.metric,
         ProcessorEvent.error,
       ],
     },
-    body: {
-      track_total_hits: false,
-      size: 0,
-      query: {
-        bool: {
-          filter: [
-            ...rangeQuery(start, end),
-            ...termQuery(SERVICE_NAME, serviceName),
-          ],
-        },
-      },
-      aggs: {
-        environments: {
-          terms: {
-            field: SERVICE_ENVIRONMENT,
-            missing: ENVIRONMENT_NOT_DEFINED.value,
-            size,
-          },
-        },
```

Comment:
We should only use terms enum API for the service inventory page. It looks like this changes the behaviour for all pages that calls the environments API.

@dgieselaar is adding this method in his PR. We might want something similar for environments:
https://github.com/elastic/kibana/blob/f0a7e457fd95ee663829d4fb64055777cfc1889d/x-pack/plugins/apm/server/routes/services/get_services/get_service_names_from_terms_enum.ts

We also have a more flexible endpoint for getting suggestions for the terms enum API. You might want to use that instead:
https://github.com/elastic/kibana/blob/f0a7e457fd95ee663829d4fb64055777cfc1889d/x-pack/plugins/apm/server/routes/suggestions/get_suggestions_with_terms_enum.ts#L11-L53

Please make sure that we have API tests that cover getting environments via the terms enum API and via the non terms API. You'll likely need to create a new synthtrace scenario to test these two scenarios.

Reply:
@sqren The second point around using the `suggestions` route is really great. I will re-use this function.

Third point, yes, i will add some API tests and create new Synthtrace. That's a good idea

For the 1st point - Regarding using the terms enum api only for the service inventory page, i am not clear with that part ? Should we create a complete new route for this which will only be called by the Service Inventory page ? Also can you provide more details as to why should we avoid the terms enum api around other calls ?

Reply:
After our discussion, what i can understood is instead of looking at this challenge from this perspective that we should only use `term_enum` api for `service inventory` page, i would rather look at this as we should only use `terms_enum` api when we don't have a service name passed. 

When we have service name passed, we want the expensive but exact result, hence we will use the existing aggregation.
When we don't have service name, then we can rely on terms_enum api.

If the above is correct, we can make a simple change here to solve this 
https://github.com/elastic/kibana/blob/ad1cd3e4e325bebd2b878845729df7aba40f4766/x-pack/plugins/apm/public/hooks/use_environments_fetcher.tsx#L12-L38